### PR TITLE
ICS8: Update the WASM Client spec

### PR DESCRIPTION
**Pull request by blasrodri**

### CODE REVIEW

File: spec/client/ics-008-wasm-client/README.md [hunk 1]
```diff
@@ -203,124 +266,207 @@ function upgradeClientState(
...
-  clientState: ClientState,
-  height: Height,
-  prefix: CommitmentPrefix,
-  proof: CommitmentProof,
-  portIdentifier: Identifier,
-  channelIdentifier: Identifier,
-  channelEnd: ChannelEnd) {
-    codeHandle = clientState.codeHandle()
-    assert(codeHandle.verifyChannelState(clientState, height, prefix, proof, portIdentifier, channelIdentifier, channelEnd))
+func (c ClientState) CheckForMisbehaviour(ctx sdk.Context, cdc codec.BinaryCodec, clientStore sdk.KVStore, msg exported.ClientMessage) bool {
```
Comment (blasrodri) [inline spec/client/ics-008-wasm-client/README.md:330]: this has to be in typescript @edjroz

File: spec/client/ics-008-wasm-client/README.md [hunk 2]
```diff
@@ -84,329 +111,457 @@ interface Height {
...
+  payload = verifyClientMessageMsg{clientMsg}
 
-    store = getStore("clients/{identifier}")
-    return codeHandle.initialise(store, initializationData, consensusState)
+  clientState = provableStore.get("clients/{clientMsg.identifier}/clientState")
+  // retrieve client identifier-prefixed store
+  clientStore = provableStore.prefixStore("clients/{clientMsg.identifier}")
+
+  // use underlying wasm contract to verify client message
+  assert(callContract(clientStore, clientState, marshalJSON(payload)))
```
Comment (aeryz) [inline spec/client/ics-008-wasm-client/README.md:200]: NIT: Since the read-only operations are now changed to be queries, I would use `queryContract` when a read-only query is made to the underlying contract and `callContract` when an execute/sudo entrypoint is called.

Comment (crodriguezvega) [reply spec/client/ics-008-wasm-client/README.md:200]: Thanks for the feedback. Yes, I agree; I will change that.

File: spec/client/ics-008-wasm-client/README.md [hunk 3]
```diff
@@ -84,329 +111,457 @@ interface Height {
...
-    assert(codeHandle.verifyNextSequenceRecv(clientState, height, prefix, proof, portIdentifier, channelIdentifier, nextSequenceRecv))
+#### Contract instantiation
+
+Instantiation of a contract is minimal. No data is passed in the message for the contract call, but the
+Wasm client store is passed. This allows for a Wasm contract to initialize any metadata that they
+need such as processed height and/or processed time.
+
+#### Contract query
+
+Every Wasm contract must to support these query messages:
```
Comment (AdityaSripal) [inline spec/client/ics-008-wasm-client/README.md:410]:
```suggestion
Every Wasm contract must support these query messages:
```

File: spec/client/ics-008-wasm-client/README.md [hunk 4]
```diff
@@ -84,329 +111,457 @@ interface Height {
...
-  height: Height,
-  prefix: CommitmentPrefix,
-  proof: CommitmentProof,
-  portIdentifier: Identifier,
-  channelIdentifier: Identifier,
-  sequence: uint64) {
-    codeHandle = clientState.codeHandle()
-    assert(codeHandle.verifyPacketReceiptAbsence(clientState, height, prefix, proof, portIdentifier, channelIdentifier, sequence))
-}
+When an instruction needs to be executed in Wasm code, functions are executed using a `wasmvm`.
```
Comment (AdityaSripal) [inline spec/client/ics-008-wasm-client/README.md:397]: Should we include that this wasmvm is sandboxed from any other vms users might write to for other purposes?

**Review by AdityaSripal — COMMENTED** [review]
Nice work! Couple minor issues also pending change to queryContract for reads as mentioned by other reviewers

File: spec/client/ics-008-wasm-client/README.md (hunk 4 above)
Comment (blasrodri) [reply spec/client/ics-008-wasm-client/README.md:397]: Added a comment

File: spec/client/ics-008-wasm-client/README.md [hunk 5]
```diff
@@ -84,329 +111,457 @@ interface Height {
...
+    height,
+    delayTimePeriod,
+    delayBlockPeriod,
+    proof,
+    path,
+    value
+  }
+
+  // retrieve client identifier-prefixed store
+	clientStore = provableStore.prefixStore("clients/{clientIdentifier}")
```
[the remark this answers is not in the document]
Comment (blasrodri) [reply spec/client/ics-008-wasm-client/README.md:359]: How should it be?

**Review by AdityaSripal — APPROVED** [review]
Since this is correct and only pending minor improvements, I'm approving and merging so we can tackle future issues in smaller PRs. THanks everyone for their hard work and patience!

File: spec/client/ics-008-wasm-client/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (DimitrisJim) [reply spec/client/ics-008-wasm-client/README.md:359]: the assignment to clientStore appears to be indented? (maybe this is githubs UI being all weird)

File: spec/client/ics-008-wasm-client/README.md [hunk 6]
```diff
@@ -84,329 +111,450 @@ interface Height {
...
-  prefix: CommitmentPrefix,
-  proof: CommitmentProof,
-  portIdentifier: Identifier,
-  channelIdentifier: Identifier,
-  nextSequenceRecv: uint64) {
-    codeHandle = clientState.codeHandle()
-    assert(codeHandle.verifyNextSequenceRecv(clientState, height, prefix, proof, portIdentifier, channelIdentifier, nextSequenceRecv))
+#### Contract instantiation
+
+Instantiation of a contract is minimal. No data is passed in the message for the contract call, but the
```
Comment (DimitrisJim) [inline spec/client/ics-008-wasm-client/README.md:405]: consensus & client state is also passed in?

File: spec/client/ics-008-wasm-client/README.md [hunk 7]
```diff
@@ -84,329 +111,450 @@ interface Height {
...
-#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
-#[serde(rename_all = "snake_case")]
-pub enum QueryMsg {
-   IsValid(ValidityPredicate),
-   LatestClientHeight(Vec<u8>),
+```rust
+#[cw_serde]
+pub struct ContractResult {
+  pub is_valid: bool, // Execute was successful
+  pub error_msg: String, // Error message
```
Comment (DimitrisJim) [inline spec/client/ics-008-wasm-client/README.md:523]: this was brought up in slack discussion at some point, I remember both `is_valid` and `error_msg` being unused in contracts. Either way, can be left for another PR after we actually remove em from code.

**Review by crodriguezvega — APPROVED** [review]
Thank you for all the work on updating the spec and the feedback from the reviewers. I will merge the PR now as is and continue improving it and addressing the remaining feedback in a separate PR.
